File: 0047_permutationsII/main.cpp
#include <vector>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
class Solution {
private:
    void dfs(std::vector<int>& perm, std::unordered_map<int,int>& countMap, std::vector<int> nums, std::vector<std::vector<int>>& res) {
        if(perm.size() == nums.size()) {
            res.push_back(perm);
            return;
        }

        for(auto [key, value]: countMap) {
            if(value == 0)
                continue;
            perm.push_back(key);
            --countMap[key];
            dfs(perm, countMap, nums, res);
            ++countMap[key];
            perm.pop_back();
        }
    }
public:
    std::vector<std::vector<int>> permuteUnique(std::vector<int>& nums) {
        std::sort(nums.begin(), nums.end());
        std::vector<std::vector<int>> res;
        std::vector<int> perm;
        std::unordered_map<int, int> countMap;
        for(auto num: nums)
            ++countMap[num];
        dfs(perm, countMap, nums, res);
        return res;
    }
};

class Solution1 {
private:
    void dfs(int startIdx, std::vector<int> nums, std::vector<std::vector<int>>& res) {
        if(startIdx == nums.size()) {
            res.push_back(nums);
            return;
        }

        std::unordered_set<int> mySet;
        for(int i = startIdx; i < nums.size(); ++i) {
            if(mySet.find(nums[i]) != mySet.end())
                continue;
            mySet.insert(nums[i]);
            std::swap(nums[startIdx], nums[i]);
            dfs(startIdx + 1, nums, res);
            std::swap(nums[startIdx], nums[i]);
        }
    }
public:
    std::vector<std::vector<int>> permuteUnique(std::vector<int>& nums) {
        std::vector<std::vector<int>> res;
        dfs(0, nums, res);
        return res;
    }
};

File: 0047_permutationsII/main.py
from typing import List
from collections import Counter

class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        res = []
        perm = []
        countMap = { num: 0 for num in nums }
        for num in nums:
            countMap[num] += 1

        def backtrack():
            if len(perm) == len(nums):
                res.append(perm.copy())
                return
            
            for num in countMap:
                if countMap[num] > 0:
                    perm.append(num)
                    countMap[num] -= 1
                    backtrack()
                    countMap[num] += 1
                    perm.pop()
        
        backtrack()
        return res

class Solution1:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        res = []
        counter = Counter(nums)
        def backtrack(perm):
            if len(perm) == len(nums):
                res.append(perm)
                return
            for key in counter:
                if counter[key]:
                    counter[key] -= 1
                    backtrack(perm + [key])
                    counter[key] += 1
        backtrack([])
        return res

Solution1().permuteUnique([1, 1, 2])